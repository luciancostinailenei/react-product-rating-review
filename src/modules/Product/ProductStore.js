import { action, computed, observable } from 'mobx'

import moment from 'moment'

class ProductStore {
  @observable reviews = []
  @observable isProductReviewFormDisplayed = false

  @action
  addReview = review => {
    this.reviews.push(review)
  }

  @computed
  get reviewsCount() {
    return this.reviews.length
  }

  @computed
  get sortedReviews() {
    const sortedReviews = this.reviews.sort((a, b) => {
      if (moment(a.date) > moment(b.date)) {
        return -1
      }

      if (moment(a.date) < moment(b.date)) {
        return 1
      }

      return 0
    })

    return sortedReviews
  }

  @computed
  get averageScore() {
    const numberOfReviews = this.reviewsCount

    if (numberOfReviews === 0) {
      return 0
    }

    const scoreSum = this.reviews.reduce((acc, review) => {
      return acc + review.rating
    }, 0)

    const averageScore = scoreSum / numberOfReviews
    return averageScore % 1 === 0 ? averageScore : averageScore.toFixed(1)
  }
}

const store = new ProductStore()
export default store
