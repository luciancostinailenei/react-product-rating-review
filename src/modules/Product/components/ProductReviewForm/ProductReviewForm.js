import React, { Component } from 'react'
import { inject } from 'mobx-react'

import './ProductReviewForm.scss'

import ProductRating from '../ProductRating/ProductRating'

@inject('ProductStore')
export default class ProductReviewForm extends Component {
  constructor(props) {
    super(props)

    this.state = this.getInitialState()

    this.submitReview = this.submitReview.bind(this)
    this.setReviewTitle = this.setReviewTitle.bind(this)
    this.setReviewText = this.setReviewText.bind(this)
    this.changeRating = this.changeRating.bind(this)
  }

  getInitialState() {
    return {
      rating: 0,
      title: '',
      text: ''
    }
  }

  areInputsFilled() {
    const { rating, title, text } = this.state
    return rating !== 0 && title !== '' && text !== ''
  }

  submitReview() {
    if (!this.areInputsFilled()) {
      return
    }

    const productReview = {
      ...this.state,
      user: 'Lucian',
      date: new Date()
    }

    this.props.ProductStore.addReview(productReview)

    this.setState({ rating: 0, title: '', text: 'Review' })
    this.refs.ratingInput.wrappedInstance.resetVote()
    this.props.ProductStore.isProductReviewFormDisplayed = false
  }

  setReviewTitle(event) {
    this.setState({ title: event.target.value })
  }

  setReviewText(event) {
    this.setState({ text: event.target.value })
  }

  changeRating(value) {
    this.setState({ rating: value })
  }

  render() {
    return (
      <div className="ProductReviewForm">
        <div className="ProductReviewForm__rating">
          <span>Rate product: </span>

          <ProductRating
            ref="ratingInput"
            inputIdIndex="100"
            changeRatingHandler={this.changeRating}
          />
        </div>

        <div className="ProductReviewForm__title">
          <input
            onChange={this.setReviewTitle}
            value={this.state.title}
            type="text"
            placeholder="Review title"
          />
        </div>

        <div className="ProductReviewForm__text">
          <textarea
            placeholder="Review..."
            value={this.state.text}
            onChange={this.setReviewText}
          />
        </div>

        <div className="ProductReviewForm__submit">
          <button
            className={
              'btn btn--primary' +
              (!this.areInputsFilled() ? ' btn--disabled' : '')
            }
            onClick={this.submitReview}
            disabled={!this.areInputsFilled()}
          >
            Submit
          </button>
          <span>Rating and all inputs are required.</span>
        </div>
      </div>
    )
  }
}
