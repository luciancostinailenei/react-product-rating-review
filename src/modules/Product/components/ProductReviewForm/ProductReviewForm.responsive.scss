/*------------------------------------*\
  #ProductReviewForm-Responsive
\*------------------------------------*/

// Small devices (landscape phones, 576px and up)
@media (max-width: 576px) {  
    .ProductReviewForm { }
    
        .ProductReviewForm__submit {
            span {
                display: block;
                margin-left: 5px;
                margin-top: 10px;
            }
        }
}

// Medium devices (tablets, 768px and up)
@media (max-width: 768px) { }

// Large devices (desktops, 992px and up)
@media (max-width: 992px) {  
    .ProductReviewForm { }

        .ProductReviewForm__text {
            textarea {
                width: 90%;
            }
        }

        .ProductReviewForm__title {
            input {
                width: 90%;
            }
        }
}

// Extra large devices (large desktops, 1200px and up)
@media (max-width: 1200px) {  }