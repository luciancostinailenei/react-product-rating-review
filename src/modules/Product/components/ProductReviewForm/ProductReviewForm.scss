.ProductReviewForm { 
    background-color: var(--teal-50);
    border-radius: 3px;
    padding: 20px;
}

    .ProductReviewForm__rating {
        color: var(--teal-700);
        align-items: center;
        display: flex;
    }

    .ProductReviewForm__rating-stars {
        margin-left: 10px;
    }
    
    .ProductReviewForm__title {
        margin-top: 10px;
        input {
            width: 500px;
            height: 20px;
            border: none;
            padding: 10px;
            color: var(--gray-600);
            border-radius: 3px;
        }
    }

    .ProductReviewForm__text {
        margin-top: 10px;

        textarea {
            width: 800px;
            height: 150px;
            border: none;
            padding: 10px;
            color: var(--gray-600);
            border-radius: 3px;
        }
    }

    .ProductReviewForm__submit {
        margin-top: 10px;
        
        span {
            color: var(--gray-700);
            font-size: 12px;
            margin-left: 10px;
        }
    }

.ProductReviewForm--hidden {
    display: none;
}