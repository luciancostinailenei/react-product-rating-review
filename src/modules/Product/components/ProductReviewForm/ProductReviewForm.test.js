import { configure, shallow } from 'enzyme'
import React from 'react'
import Adapter from 'enzyme-adapter-react-16'
import toJson, { createSerializer } from 'enzyme-to-json'

import ProductReviewForm from './ProductReviewForm'

import ProductStore from '../../ProductStore'

expect.addSnapshotSerializer(createSerializer({ mode: 'deep' }))

configure({ adapter: new Adapter() })

describe('ProductReviewForm', () => {
  it('matches snapshot', () => {
    const ProductReviewsWrapper = shallow(
      <ProductReviewForm ProductStore={ProductStore} />
    )

    expect(toJson(ProductReviewsWrapper)).toMatchSnapshot()
  })
})
