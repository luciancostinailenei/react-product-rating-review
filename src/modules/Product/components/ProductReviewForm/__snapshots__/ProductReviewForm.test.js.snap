// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductReviewForm matches snapshot 1`] = `
<ProductReviewForm
  ProductStore={ProductStore {}}
/>
`;
